glean-core/src/upload/mod.rs

```diff
@@ -30,49 +30,82 @@ use util::process_pings_dir;
 pub struct PingUploadManager {
     /// A FIFO queue storing a `PingRequest` for each pending ping.
     queue: Arc<RwLock<VecDeque<PingRequest>>>,
-    /// A flag, signaling if we are done processing the pending pings directories.
-    processed_pending_pings: Arc<AtomicBool>,
+    /// A flag signaling if we are currently processing the pending pings directories.
+    processing_pending_pings: Arc<AtomicBool>,
 }
 
 impl PingUploadManager {
     /// Create a new PingUploadManager.
-    ///
-    /// This will spawn a new thread and processes the pending pings folder
-    /// filling up the queue with whatever pings are in there
-    /// ordered by oldest to newest file modified.
-    pub fn new(data_path: &str) -> Self {
-        let queue = Arc::new(RwLock::new(VecDeque::new()));
-        let processed_pending_pings = Arc::new(AtomicBool::new(false));
-
-        let data_path = PathBuf::from_str(&data_path).expect("data_path must be a valid path.");
-        let local_queue = queue.clone();
-        let local_flag = processed_pending_pings.clone();
-        let _ = thread::Builder::new()
-            .name("glean.upload_manager.process_pings_directory".to_string())
-            .spawn(move || match process_pings_dir(&data_path) {
-                Ok(requests) => {
-                    let mut local_queue = local_queue
-                        .write()
-                        .expect("Can't write to pending pings queue.");
-                    local_queue.extend(requests.into_iter());
-                    local_flag.store(true, Ordering::SeqCst);
-                }
-                Err(e) => log::info!("Error processing pending pings directories! {}", e),
-            });
-
+    pub fn new() -> Self {
         Self {
-            queue,
-            processed_pending_pings,
+            queue: Arc::new(RwLock::new(VecDeque::new())),
+            processing_pending_pings: Arc::new(AtomicBool::new(false)),
         }
     }
 
-    fn has_processed_pending_pings(&self) -> bool {
-        self.processed_pending_pings.load(Ordering::SeqCst)
+    fn is_processing_pending_pings(&self) -> bool {
+        self.processing_pending_pings.load(Ordering::SeqCst)
+    }
+
+    /// Spawns a new thread and processes the pending pings directory,
+    /// filling up the queue with whatever pings are in there.
+    ///
+    /// While doing the processing, the processing_pending_pings flag will be set to `true`.
+    /// It will be turned back to `false` once the processing is done,
+    /// no matter if it was successfull or not.
+    ///
+    /// # Arguments
+    ///
+    /// * `data_path` - Path to the folder where the pending pings directories are.
+    ///
+    /// # Return value
+    ///
+    /// `thread::JoinHandle`, the spawned thread handle (for tests synchronization purposes).
+    ///
+    /// # Panics
+    ///
+    /// Will panic if unable to spawn a new thread.
+    pub fn process_pending_pings(&self, data_path: &str) -> thread::JoinHandle<()> {
```

Comment:
Most of the changes I made to this function, even the change where I remove it from `::new` into it's own function, were so that I could test this in an easier manner. These changes don't however change how it used to work, apart from:
1)  Now it panicks in case not able to spawn a new thread, I just thought that made sense;
2) It will change the flag even when not sucessful processing the directory. I also just thought that made sense.

Reply:
Seems like good choices.

Reply:
Yes, both are good ideas! Please comment in the code on *why*, though. Especially why you're changing the flag even though processing has failed.